File: src/storage/disk.ts
/**
 * Local filesystem storage backend.
 */
import { mkdir, readdir, unlink, stat } from "node:fs/promises";
import { join, dirname, resolve } from "node:path";
import type { ReadableStream } from "node:stream/web";
import type { StorageBackend } from "./backend.js";

export class DiskStorage implements StorageBackend {
  private basePath: string;

  constructor(basePath: string) {
    this.basePath = resolve(basePath);
  }

  private resolve(key: string): string {
    return join(this.basePath, key);
  }

  async write(key: string, data: Uint8Array | string): Promise<void> {
    const fullPath = this.resolve(key);
    await mkdir(dirname(fullPath), { recursive: true });
    await Bun.write(fullPath, data);
  }

  async read(key: string): Promise<Uint8Array> {
    const file = Bun.file(this.resolve(key));
    const buf = await file.arrayBuffer();
    return new Uint8Array(buf);
  }

  readStream(key: string): ReadableStream<Uint8Array> {
    const file = Bun.file(this.resolve(key));
    return file.stream() as unknown as ReadableStream<Uint8Array>;
  }

  async list(prefix: string): Promise<string[]> {
    const prefixPath = this.resolve(prefix);
    try {
      const s = await stat(prefixPath);
      if (s.isFile()) return [prefix];
    } catch {
      return [];
    }

    const keys: string[] = [];
    const walk = async (dir: string): Promise<void> => {
      const entries = await readdir(dir, { withFileTypes: true });
      for (const entry of entries) {
        const full = join(dir, entry.name);
        if (entry.isDirectory()) {
          await walk(full);
        } else if (entry.isFile()) {
          // Make key relative to basePath
          keys.push(full.slice(this.basePath.length + 1));
        }
      }
    };

    await walk(prefixPath);
    return keys.sort();
  }

  async exists(key: string): Promise<boolean> {
    const file = Bun.file(this.resolve(key));
    return file.exists();
  }

  async delete(key: string): Promise<void> {
    try {
      await unlink(this.resolve(key));
    } catch {
      // Ignore if not found
    }
  }
}

File: context_use/storage/disk.py
from __future__ import annotations

from pathlib import Path
from typing import BinaryIO

from context_use.storage.base import StorageBackend


class DiskStorage(StorageBackend):
    """Local filesystem storage backend."""

    def __init__(self, base_path: str) -> None:
        self._base = Path(base_path)
        self._base.mkdir(parents=True, exist_ok=True)

    def _resolve(self, key: str) -> Path:
        return self._base / key

    # ---- interface ----

    def write(self, key: str, data: bytes) -> None:
        path = self._resolve(key)
        path.parent.mkdir(parents=True, exist_ok=True)
        path.write_bytes(data)

    def read(self, key: str) -> bytes:
        return self._resolve(key).read_bytes()

    def open_stream(self, key: str) -> BinaryIO:
        path = self._resolve(key)
        return open(path, "rb")  # noqa: SIM115

    def list_keys(self, prefix: str) -> list[str]:
        prefix_path = self._resolve(prefix)
        if not prefix_path.exists():
            return []
        if prefix_path.is_file():
            return [prefix]
        keys: list[str] = []
        for p in prefix_path.rglob("*"):
            if p.is_file():
                keys.append(str(p.relative_to(self._base)))
        return sorted(keys)

    def exists(self, key: str) -> bool:
        return self._resolve(key).exists()

    def delete(self, key: str) -> None:
        path = self._resolve(key)
        if path.is_file():
            path.unlink()
